compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -2838,26 +2841,271 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
-                        err.span_label(
-                            span,
-                            if ty.references_error() {
-                                String::new()
-                            } else {
-                                format!("this tail expression is of type `{:?}`", ty)
-                            },
+                    if let hir::Expr { kind: hir::ExprKind::Block(..), .. } = expr {
+                        let expr = expr.peel_blocks();
+                        let ty =
+                            typeck_results.expr_ty_adjusted_opt(expr).unwrap_or(tcx.ty_error());
+                        let span = expr.span;
+                        if Some(span) != err.span.primary_span() {
+                            err.span_label(
+                                span,
+                                if ty.references_error() {
+                                    String::new()
+                                } else {
+                                    let ty = with_forced_trimmed_paths!(self.ty_to_string(ty));
+                                    format!("this tail expression is of type `{ty}`")
+                                },
+                            );
+                        }
+                    }
+
+                    let mut primary_spans = vec![];
+                    let mut span_labels = vec![];
+
+                    // FIXME: visit the ty to see if there's any closure involved, and if there is,
+                    // check whether its evaluated return type is the same as the one corresponding
+                    // to an associated type (as seen from `trait_pred`) in the predicate. Like in
+                    // trait_pred `S: Sum<<Self as Iterator>::Item>` and predicate `i32: Sum<&()>`
+                    let mut type_diffs = vec![];
+
+                    if let ObligationCauseCode::ExprBindingObligation(def_id, _, _, idx) = parent_code.deref()
+                        && let predicates = self.tcx.predicates_of(def_id).instantiate_identity(self.tcx)
+                        && let Some(pred) = predicates.predicates.get(*idx)
+                        && let ty::PredicateKind::Clause(ty::Clause::Trait(trait_pred)) = pred.kind().skip_binder()
+                    {
+                        let mut c = CollectAllMismatches {
+                            infcx: self.infcx,
+                            param_env: param_env,
+                            errors: vec![],
+                        };
+                        if let ty::PredicateKind::Clause(ty::Clause::Trait(
+                                predicate
+                            )) = predicate.kind().skip_binder()
+                        {
+                            if let Ok(_) = c.relate(trait_pred, predicate) {
+                                type_diffs = c.errors;
+                            }
+                        }
+                    }
+                    let point_at_chain = |expr: &hir::Expr<'_>| {
```

Comment:
This should be a separate function. Way too long to be called later.

Reply:
Moved to its own method, I'll file a follow up PR to split the trait through multiple files to make it more manageable.